### PR TITLE
370 - vm_clone produces VMs that are silently fragile to multi-disk attach

### LINKED ISSUE #370: vm_clone produces VMs that are silently fragile to multi-disk attach + reboot (no bootDevices, vda-anchored grub)
## Summary

VMs cloned via `scale_computing.hypercore.vm_clone` from a standard Canonical Ubuntu cloud image inherit two independent boot-fragility issues that surface the first time a second `VIRTIO_DISK` is attached AND the VM is rebooted:

1. **`bootDevices` is empty** on the cloned VirDomain. HC's BIOS has an implicit auto-fallback to "the only virtio disk" but that fallback gives up once a second virtio is attached → `Boot failed: not a bootable disk / No bootable device`.
2. **The cloud image's `/etc/default/grub` ships with `GRUB_DISABLE_LINUX_UUID=true`**, so the generated `/boot/grub/grub.cfg` uses `root=/dev/vda1` (device-path) instead of `root=UUID=…` (UUID-based). When a second virtio disk reorders PCI enumeration, the OS disk becomes `vdb`, the kernel can't find `/dev/vda1`, and initramfs hangs at `Btrfs loaded …`.

Combined effect: a cloud-image VM cloned via this module **boots correctly for years** while it has exactly one virtio disk, then bricks on the first reboot after any tool (CSI driver, manual disk-add, Terraform provider, or even another playbook in the same suite) attaches a second virtio disk.

This is the underlying root cause for a pair of related issues filed in `ScaleComputing/k3s-on-hypercore` (#7 and #8), but the same exposure exists in **every** downstream consumer of `vm_clone`: `k3s-ansible-hypercore`, `ansible_edge_playbooks`, and customer playbooks that clone from cloud-image templates.

## Reproducer

```bash
# Clone an Ubuntu cloud-image VM via this collection (any of the standard
# patterns, e.g. ansible_edge_playbooks/simple_vm_deploy.yml):
ansible-playbook simple_vm_deploy.yml

# Inspect the resulting VM:
curl -sk -u admin:admin "https://<hc-host>/rest/v1/VirDomain/<new-vm-uuid>" \
  | jq '.[0] | {name, bootDevices}'
# → bootDevices: []   ← issue #1

# SSH into the new VM:
grep GRUB_DISABLE_LINUX_UUID /etc/default/grub
# → GRUB_DISABLE_LINUX_UUID=true     ← issue #2
cat /proc/cmdline
# → root=/dev/vda1    ← issue #2 consequence

# Attach a second VIRTIO_DISK, then reboot:
#   - If issue #1 unaddressed: BIOS "No bootable device" — VM unbootable.
#   - If issue #1 fixed but #2 unaddressed: kernel loads but initramfs hangs.
```

### Layer 1: `vm_clone` should set `bootDevices` after clone

The most natural fix is to add a parameter (or change defaults) so the cloned VM's `bootDevices` is populated with the primary disk's UUID immediately after creation. Something like:

```yaml
- name: Clone the VM
  scale_computing.hypercore.vm_clone:
    vm_name: "{{ inventory_hostname }}"
    source_vm_name: "{{ source_template }}"
    set_boot_devices: yes   # new default-yes parameter
```

Or simpler: always populate `bootDevices` with the cloned VM's largest VIRTIO_DISK by default. Users who want to manage boot order manually can override.

### Layer 2: optional `cloud_init.runcmd` injection

If the module accepts `cloud_init.user_data`, the generated cloud-init should include a `runcmd` to fix the grub UUID issue at first boot:

```yaml
runcmd:
  - sed -i 's/^GRUB_DISABLE_LINUX_UUID=true/#GRUB_DISABLE_LINUX_UUID=true/' /etc/default/grub
  - update-grub
```

Either bake this into the module's default user_data, or document it prominently in the role's README so every downstream playbook can add it.

## Workaround for the field today

I've documented both layers + per-VM fix recipes in an internal `hypercore-api-notes.md` reference. The TL;DR for any playbook today:

1. After `vm_clone`, `PATCH /VirDomain/{uuid}` with `bootDevices: [<primary-disk-uuid>]`.
2. Include the grub `runcmd` snippet in the cloud-init user data.

I've also done both layers manually on an 8-VM reference cluster — works cleanly.

## Why this matters

Without these fixes, **every k8s-on-HyperCore deployment is a ticking time bomb**. The VMs come up fine, work fine for months/years, then the first time someone adds storage (CSI driver, data disk, NFS export disk, etc.) and the VM reboots, it bricks. That's exactly the wrong failure pattern: silent and recovery-blocking.

## Related

- HC platform team issue (BIOS auto-fallback should be smarter or refuse empty `bootDevices`): `github.lab.local/dev/dev#394`
- `k3s-on-hypercore#7` (bootDevices)
- `k3s-on-hypercore#8` (grub UUID)

## Suggested labels

`bug`, `priority:high` — silent failure mode that affects every multi-disk workload on HC.

### PR DESCRIPTION
**Pull request by domendobnikar**
Boot order is now set after cloning to avoid issues when attaching additional disk to VM afterwards. 
`Cloud init's user data` default `runcmd `is also added if `cloud init` has been passed as parameter.

### CODE REVIEW

File: plugins/modules/vm_clone.py [hunk 1]
```diff
@@ -193,7 +209,14 @@ def main():
...
                 options=dict(
-                    user_data=dict(type="str"),
+                    user_data=dict(
+                        type="str",
+                        default=(
+                            "runcmd:\n"
+                            "  - sed -i 's/^GRUB_DISABLE_LINUX_UUID=true/#GRUB_DISABLE_LINUX_UUID=true/' /etc/default/grub\n"
+                            "  - update-grub\n"
+                        ),
+                    ),
```
[the remark this answers is not in the document]
Comment (domendobnikar) [reply plugins/modules/vm_clone.py]: Returned to the original module structure, added user_data "default" during payload creation.

File: plugins/module_utils/vm.py [hunk 2]
```diff
@@ -610,6 +610,10 @@ def find_disk(self, slot):
             if disk.slot == slot:
                 return disk
 
+    # primary disk is the largest Virtio disk
+    def get_primary_disk(self):
+        return max(self.disk_list, key=lambda disk: disk.size)
```
[the remark this answers is not in the document]
Comment (domendobnikar) [reply plugins/module_utils/vm.py]: Added type

**Comment by domendobnikar** [comment]
Integration tests: 
https://github.com/ScaleComputing/HyperCoreAnsibleCollection/actions/runs/27026276199/job/79767085521

File: plugins/module_utils/vm.py [hunk 3]
```diff
@@ -423,6 +423,27 @@ def create_export_or_import_vm_payload(ansible_dict, cloud_init, is_export):
...
         return payload
 
+    @staticmethod
+    def clone_add_user_data_to_cloud_init(cloud_init):
+        # Task 370 - the generated cloud-init should include a runcmd
+        # to fix the grub UUID issue at first boot
+
+        user_data = cloud_init.get("user_data")
+        if not user_data:
+            return cloud_init
```
Comment (justinc1) [inline plugins/module_utils/vm.py:433]: No need to fix grub config in this case?

File: plugins/module_utils/vm.py [hunk 4]
```diff
@@ -423,6 +423,27 @@ def create_export_or_import_vm_payload(ansible_dict, cloud_init, is_export):
...
+
+        cloud_init["user_data"] = (
+            user_data.rstrip()
+            + """
+
+        runcmd:
```
Comment (justinc1) [inline plugins/module_utils/vm.py:439]: We might end adding a user_data section when we already have a user_data section.
We can add a unit tests for those corner cases, then refaction this function a bit.
Corner cases:
- no user_data
- emtpy user_data
- no user_data.runcmd
- emtpy user_data.runcmd
- user_data.runcmd with entries

I guess we want to always comment out GRUB_DISABLE_LINUX_UUID ?

File: plugins/module_utils/vm.py [hunk 5]
```diff
@@ -446,6 +467,7 @@ def create_clone_vm_payload(
                     hypercore_tags.append(tag)
             data["template"]["tags"] = ",".join(hypercore_tags)
         if cloud_init:
+            cloud_init = cls.clone_add_user_data_to_cloud_init(cloud_init)
```
Comment (justinc1) [inline plugins/module_utils/vm.py:470]: I have impression we want to alway run this?

File: tests/integration/targets/vm_clone/tasks/10_clone_stopped.yml [hunk 6]
```diff
@@ -29,11 +29,12 @@
       - cloned_info.records | length == 1
       - source_info.records.0.vcpu == cloned_info.records.0.vcpu
       - source_info.records.0.tags != cloned_info.records.0.tags
-      - source_info.records.0.boot_devices | length == cloned_info.records.0.boot_devices | length
       - source_info.records.0.disks | length ==  cloned_info.records.0.disks | length
       - source_info.records.0.nics | length ==  cloned_info.records.0.nics | length
       - source_info.records.0.nics.0.mac !=  cloned_info.records.0.nics.0.mac
       - source_info.records.0.node_affinity == cloned_info.records.0.node_affinity
+      # Cloned VM's boot devices should be set
+      - cloned_info.records.0.boot_devices | length != 0
```
Comment (justinc1) [inline tests/integration/targets/vm_clone/tasks/10_clone_stopped.yml:37]: I would prefer to test with "==".

I guess `source_info.records.0.boot_devices | length` == 0?

And `cloned_info.records.0.boot_devices | length` was 0 before, 1 after PR?

File: plugins/module_utils/vm.py [hunk 7]
```diff
@@ -610,6 +632,13 @@ def find_disk(self, slot):
             if disk.slot == slot:
                 return disk
 
+    # primary disk is the largest Virtio disk
+    def get_primary_disk(self):
```
Comment (justinc1) [inline plugins/module_utils/vm.py:636]: We could clone a VM with IDE or SCSI disk(s).

Not sure why is the largest disk the bootable one. VM with 100 kB cloud-init ISO, 20 GB OS disk, 100 GB installed-apps disk. Maybe the first disk on bus is the OS disk, but if we add/remove disk, we likely can change that too. :( :(

Ideally we could ask HC3 about this setting for the source VM. But source VM might never be booted before clonning.

Or, we could clone VM, boot it, then assuming cloned VM does boot (really boot, not just wait on 
"no OS installed") set this from cloned VM runtime state.

A NIC could be a boot device too.

**Review by justinc1 — COMMENTED** [review]
I think PR solves exactly the problem described in issue (e.g. virtio disk etc), but also creates additional corner cases.

I think we need to investigate what can be done. What if sorce VM has IDE or SCSI disk, a mixture, a NIC boot device etc.

I'm not sure we will be able to fully solve the problem. I tried to create a simple emtpy VM, with 1 disk, in HC3 web UI. The VM got beside the disk also a NIC, (empty) cloud-init CD-ROM, and scale-guest-tools-4.2.iso CD-ROM. The disk and  (empty) cloud-init CD-ROM were set as boot devices by HC3. In this case it seems like source VM boot order could be used as source of truth. But I did not check all potential ways to create a source VM.

File: plugins/module_utils/vm.py (hunk 3 above)
Comment (domendobnikar) [reply plugins/module_utils/vm.py:433]: I'm not 100% sure, this is from the task description: "_If the module accepts cloud_init.user_data, the generated cloud-init should include a runcmd to fix the grub UUID issue at first boot_". 
I read it as "if the the user_data was passed"

File: plugins/module_utils/vm.py (hunk 7 above)
Comment (domendobnikar) [reply plugins/module_utils/vm.py:636]: _a cloud-image VM cloned via this module boots correctly for years while it has exactly one virtio disk, then bricks on the first reboot after any tool (CSI driver, manual disk-add, Terraform provider, or even another playbook in the same suite) attaches a second virtio disk._ 
From my understanding this issue is a corner case to solve the "one virtio disk already exists and is used for boot but isn't specified as one and causes issues later on when another disk is added"
